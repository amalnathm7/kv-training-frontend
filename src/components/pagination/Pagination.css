.pagination-button {
    margin-left: 10px;
    margin-right: 10px;
    border: none;
    outline: none;
    cursor: pointer;
    background-color: transparent;
}

.pagination-page-number {
    font-size: 16px;
    width: 50px;
    border: none;
}