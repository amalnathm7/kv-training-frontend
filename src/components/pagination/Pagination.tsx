import React, { useEffect } from 'react';
import './Pagination.css';
import { PAGE_LENGTH } from '../../constants/apiConstants';

export type PaginationPropsType = {
  page: number;
  length: number;
  total: number;
  setPage: React.Dispatch<React.SetStateAction<number>>;
};

const Pagination: React.FC<PaginationPropsType> = (props) => {
  const totalPages = Math.ceil(props.total / PAGE_LENGTH);

  useEffect(() => {
    if (props.total === 0) props.setPage(0);
    else if (props.page === 0) props.setPage(1);
  }, [props.total, props.page]);

  const onChangePage = (event) => {
    props.setPage(event.target.value);
  };

  const setPageValue = (value) => {
    if (value === '') props.setPage(1);
    else if (value > totalPages) props.setPage(totalPages);
    else if (value < 1) props.setPage(1);
    else props.setPage(Number(value));
  };

  const onBlur = (event) => {
    setPageValue(event.target.value);
  };

  const onKeyUp = (event) => {
    if (event.code === 'Enter') {
      setPageValue(event.target.value);
      event.target.blur();
    }
  };

  const onNextClicked = () => {
    props.setPage(props.page + 1);
  };

  const onBackClicked = () => {
    props.setPage(props.page - 1);
  };

  return (
    <td>
      {typeof props.page !== 'string' && props.page > 1 && (
        <button
          onClick={onBackClicked}
          className='pagination-button'
          data-testid='pagination-back-test'
        >
          <u>Back</u>
        </button>
      )}
      <label style={{ marginRight: '10px' }}>{`Showing ${props.length} entries in page`}</label>
      <input
        type='number'
        onBlur={onBlur}
        onKeyUp={onKeyUp}
        className='pagination-page-number'
        value={props.page}
        onChange={onChangePage}
        data-testid='pagination-page-number-test'
      ></input>
      <label>{` of ${totalPages}`}</label>
      {typeof props.page !== 'string' && props.page < totalPages && (
        <button
          onClick={onNextClicked}
          className='pagination-button'
          data-testid='pagination-next-test'
        >
          <u>Next</u>
        </button>
      )}
    </td>
  );
};

export default Pagination;
