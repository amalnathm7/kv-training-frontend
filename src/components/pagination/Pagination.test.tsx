import Pagination, { PaginationPropsType } from './Pagination';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

describe('Pagination Props Test', () => {
  test('If pagination rendered correctly', () => {
    const props: PaginationPropsType = {
      page: 1,
      length: 10,
      total: 25,
      setPage: () => {}
    };

    const element = render(
      <table>
        <tbody>
          <tr>
            <Pagination {...props} />
          </tr>
        </tbody>
      </table>
    );

    expect(element).toMatchSnapshot();
  });
  test('If pagination rendered correctly after page was set 0 and total became greater than 0', () => {
    const props: PaginationPropsType = {
      page: 0,
      length: 10,
      total: 25,
      setPage: () => {}
    };

    const element = render(
      <table>
        <tbody>
          <tr>
            <Pagination {...props} />
          </tr>
        </tbody>
      </table>
    );

    expect(element).toMatchSnapshot();
  });

  test('If pagination next button clicks', () => {
    const setPage = jest.fn((value: number) => ({ value }));

    const props: PaginationPropsType = {
      page: 1,
      length: 10,
      total: 25,
      setPage
    };

    render(
      <table>
        <tbody>
          <tr>
            <Pagination {...props} />
          </tr>
        </tbody>
      </table>
    );
    const element = screen.getByTestId('pagination-next-test');

    element.click();
    expect(setPage).toBeCalled();
  });

  test('If pagination back button clicks', () => {
    const setPage = jest.fn((value: number) => ({ value }));

    const props: PaginationPropsType = {
      page: 2,
      length: 10,
      total: 25,
      setPage
    };

    render(
      <table>
        <tbody>
          <tr>
            <Pagination {...props} />
          </tr>
        </tbody>
      </table>
    );
    const element = screen.getByTestId('pagination-back-test');

    element.click();
    expect(setPage).toBeCalled();
  });

  test('If pagination page input changes', () => {
    const setPage = jest.fn((value: number) => ({ value }));

    const props: PaginationPropsType = {
      page: 1,
      length: 10,
      total: 25,
      setPage
    };

    render(
      <table>
        <tbody>
          <tr>
            <Pagination {...props} />
          </tr>
        </tbody>
      </table>
    );
    const element = screen.getByTestId('pagination-page-number-test');

    fireEvent.change(element, { target: { value: '2' } });
    expect(setPage).toBeCalled();
  });

  test('If pagination page input changes with no value', () => {
    const setPage = jest.fn((value: number) => ({ value }));

    const props: PaginationPropsType = {
      page: 1,
      length: 10,
      total: 25,
      setPage
    };

    render(
      <table>
        <tbody>
          <tr>
            <Pagination {...props} />
          </tr>
        </tbody>
      </table>
    );
    const element = screen.getByTestId('pagination-page-number-test');

    fireEvent.blur(element, { target: { value: '' } });
    expect(setPage).toBeCalled();
  });

  test('If pagination page input changes with value greater than totalPages', () => {
    const setPage = jest.fn((value: number) => ({ value }));

    const props: PaginationPropsType = {
      page: 1,
      length: 10,
      total: 25,
      setPage
    };

    render(
      <table>
        <tbody>
          <tr>
            <Pagination {...props} />
          </tr>
        </tbody>
      </table>
    );
    const element = screen.getByTestId('pagination-page-number-test');

    fireEvent.blur(element, { target: { value: 100 } });
    expect(setPage).toBeCalled();
  });

  test('If pagination page input changes with value less than 1', () => {
    const setPage = jest.fn((value: number) => ({ value }));

    const props: PaginationPropsType = {
      page: 1,
      length: 10,
      total: 25,
      setPage
    };

    render(
      <table>
        <tbody>
          <tr>
            <Pagination {...props} />
          </tr>
        </tbody>
      </table>
    );
    const element = screen.getByTestId('pagination-page-number-test');

    fireEvent.blur(element, { target: { value: -100 } });
    expect(setPage).toBeCalled();
  });

  test('If pagination page input blurs', () => {
    const setPage = jest.fn((value: number) => ({ value }));

    const props: PaginationPropsType = {
      page: 1,
      length: 10,
      total: 25,
      setPage
    };

    render(
      <table>
        <tbody>
          <tr>
            <Pagination {...props} />
          </tr>
        </tbody>
      </table>
    );
    const element = screen.getByTestId('pagination-page-number-test');

    element.focus();
    element.blur();
    expect(setPage).toBeCalled();
  });

  test('If pagination page input pressed enter key', () => {
    const setPage = jest.fn((value: number) => ({ value }));

    const props: PaginationPropsType = {
      page: 1,
      length: 10,
      total: 25,
      setPage
    };

    render(
      <table>
        <tbody>
          <tr>
            <Pagination {...props} />
          </tr>
        </tbody>
      </table>
    );
    const element = screen.getByTestId('pagination-page-number-test');

    fireEvent.keyUp(element, { code: 'Enter' });
    expect(setPage).toBeCalled();
  });

  test('If pagination rendered correctly without list', () => {
    const setPage = jest.fn((value: number) => ({ value }));

    const props: PaginationPropsType = {
      page: 1,
      length: 0,
      total: 0,
      setPage
    };

    const element = render(
      <table>
        <tbody>
          <tr>
            <Pagination {...props} />
          </tr>
        </tbody>
      </table>
    );

    expect(element).toMatchSnapshot();
  });
});
