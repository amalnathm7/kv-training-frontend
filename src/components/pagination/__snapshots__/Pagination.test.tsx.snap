// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Props Test If pagination rendered correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table>
        <tbody>
          <tr>
            <td>
              <label
                style="margin-right: 10px;"
              >
                Showing 10 entries in page
              </label>
              <input
                class="pagination-page-number"
                data-testid="pagination-page-number-test"
                type="number"
                value="1"
              />
              <label>
                 of 3
              </label>
              <button
                class="pagination-button"
                data-testid="pagination-next-test"
              >
                <u>
                  Next
                </u>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table>
      <tbody>
        <tr>
          <td>
            <label
              style="margin-right: 10px;"
            >
              Showing 10 entries in page
            </label>
            <input
              class="pagination-page-number"
              data-testid="pagination-page-number-test"
              type="number"
              value="1"
            />
            <label>
               of 3
            </label>
            <button
              class="pagination-button"
              data-testid="pagination-next-test"
            >
              <u>
                Next
              </u>
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Pagination Props Test If pagination rendered correctly after page was set 0 and total became greater than 0 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table>
        <tbody>
          <tr>
            <td>
              <label
                style="margin-right: 10px;"
              >
                Showing 10 entries in page
              </label>
              <input
                class="pagination-page-number"
                data-testid="pagination-page-number-test"
                type="number"
                value="0"
              />
              <label>
                 of 3
              </label>
              <button
                class="pagination-button"
                data-testid="pagination-next-test"
              >
                <u>
                  Next
                </u>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table>
      <tbody>
        <tr>
          <td>
            <label
              style="margin-right: 10px;"
            >
              Showing 10 entries in page
            </label>
            <input
              class="pagination-page-number"
              data-testid="pagination-page-number-test"
              type="number"
              value="0"
            />
            <label>
               of 3
            </label>
            <button
              class="pagination-button"
              data-testid="pagination-next-test"
            >
              <u>
                Next
              </u>
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Pagination Props Test If pagination rendered correctly without list 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table>
        <tbody>
          <tr>
            <td>
              <label
                style="margin-right: 10px;"
              >
                Showing 0 entries in page
              </label>
              <input
                class="pagination-page-number"
                data-testid="pagination-page-number-test"
                type="number"
                value="1"
              />
              <label>
                 of 0
              </label>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table>
      <tbody>
        <tr>
          <td>
            <label
              style="margin-right: 10px;"
            >
              Showing 0 entries in page
            </label>
            <input
              class="pagination-page-number"
              data-testid="pagination-page-number-test"
              type="number"
              value="1"
            />
            <label>
               of 0
            </label>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
